### ISSUE TITLE
Intel v16 fails to compile C_AmrCoreLib/AmrCore.H (via Nyx)

### ISSUE THREAD

**Issue by bcfriesen** [BoxLib-Codes/BoxLib#9]
When I compile Nyx (commit aa16c5c) with BoxLib (commit 6b1063a), Intel v16 fails with the following error:

```
                 from /global/homes/f/friesen/BoxLib/Src/C_AMRLib/AmrLevel.H(12),
                 from ../../Source/Nyx.H(6),
                 from ../../Source/advance_particles.cpp(4):
/global/homes/f/friesen/BoxLib/Src/C_AmrCoreLib/AmrCore.H(115): error: qualified name is not allowed
      std::unique_ptr<AmrParGDB> m_gdb;
      ^

In file included from /global/homes/f/friesen/BoxLib/Src/C_AMRLib/Amr.H(14),
                 from /global/homes/f/friesen/BoxLib/Src/C_AMRLib/AmrLevel.H(12),
                 from ../../Source/Nyx.H(6),
                 from ../../Source/advance_particles.cpp(4):
/global/homes/f/friesen/BoxLib/Src/C_AmrCoreLib/AmrCore.H(115): error #77: this declaration has no storage class or type specifier
      std::unique_ptr<AmrParGDB> m_gdb;
      ^

In file included from /global/homes/f/friesen/BoxLib/Src/C_AMRLib/Amr.H(14),
                 from /global/homes/f/friesen/BoxLib/Src/C_AMRLib/AmrLevel.H(12),
                 from ../../Source/Nyx.H(6),
                 from ../../Source/advance_particles.cpp(4):
/global/homes/f/friesen/BoxLib/Src/C_AmrCoreLib/AmrCore.H(115): error: expected a ";"
      std::unique_ptr<AmrParGDB> m_gdb;
                     ^

In file included from /global/homes/f/friesen/BoxLib/Src/C_AMRLib/Amr.H(14),
                 from /global/homes/f/friesen/BoxLib/Src/C_AMRLib/AmrLevel.H(12),
                 from ../../Source/Nyx.H(6),
                 from ../../Source/advance_particles.cpp(4):
/global/homes/f/friesen/BoxLib/Src/C_AmrCoreLib/AmrCore.H(56): error: identifier "m_gdb" is undefined
      AmrParGDB* GetParGDB () const { return m_gdb.get(); }
```

**Comment by WeiqunZhang** [comment]
module load gcc/4.9.3.

This is a NERSC issue.  When PrgEnv-intel is loaded, a version of gcc that
works with intel's C++11 implementation should be automatically loaded.

Weiqun

**Comment by WeiqunZhang** [comment]
The reason I did not submit a ticket to NERSC is this has already been a
known issue to NERSC for about two years.

http://www.nersc.gov/users/software/compilers/intel-fortran-c-and-c/intel-bug-reports/c-11-header-files-appear-missing-on-edison/

Note the the current default of gcc is gcc/6.  I believe it does not
actually work with intel.

Weiqun

**Comment by bcfriesen** [comment]
OK, thanks for the info. I will close this issue then, since it is not a BoxLib problem.  I will also check on the status of that bug report. Scott French left NERSC a while ago so probably nobody has followed up on it recently.

What do you think of the following workaround? It causes compiling to fail unless the gcc/4.9.3 module is loaded. This may be annoying to BoxLib users at NERSC because the Intel compilers are loaded by default on Cori and Edison.

```
diff --git a/Tools/C_mk/Make.defs b/Tools/C_mk/Make.defs
index 658c5c8..2eefa84 100644
--- a/Tools/C_mk/Make.defs
+++ b/Tools/C_mk/Make.defs
@@ -490,10 +490,16 @@ ifeq ($(COMP),Intel)
    endif

    ifeq ($(WHICHLINUX), CORI)
+     ifneq ($(GCC_PATH), /opt/gcc/4.9.3)
+       $(error On Cori and Edison you must load the gcc/4.9.3 module to compile C++11 features with Intel compilers. This is a Cray bug)
+     endif
      CXXOPTF += -g -O2 -ip -qopt-report=5 -qopt-report-phase=vec -fp-model source
    endif

    ifeq ($(WHICHLINUX), EDISON)
+     ifneq ($(GCC_PATH), /opt/gcc/4.9.3)
+       $(error On Cori and Edison you must load the gcc/4.9.3 module to compile C++11 features with Intel compilers. This is a Cray bug)
+     endif
      CXXOPTF += -g -O2 -ip -qopt-report=5 -qopt-report-phase=vec -fp-model source
    endif
```

**Comment by WeiqunZhang** [comment]
I am OK with a workaround like that.  But I have two comments:

(1) It is needed only when USE_CXX11=TRUE.  Our particle class now uses
C++11, but not the other parts of BoxLib.

(2) Maybe it should also check the intel compiler versions.

Weiqun

**Comment by bcfriesen** [comment]
Thanks for the comments.

> (1) It is needed only when USE_CXX11=TRUE.  Our particle class now uses
> C++11, but not the other parts of BoxLib.

Understood.

> (2) Maybe it should also check the intel compiler versions.

Compiling fails with the same error with both v15 and v16. So the version check is not necessary.

I also checked the status of bug report with Cray. They closed it as they do not plan to fix it. They suggested adding the compiler flag `-cxxlib=/opt/gcc/default/snos` which currently points to GCC 6.1.0. However, this still fails, as the compiler still finds headers going back to GCC 4.3. I tried setting `-cxxlib=/opt/gcc/4.9.3/snos`, but that still traces back to GCC 4.3, and fails for the same reason.

Here is v2 of the patch. Please let me know what you think:

```
diff --git a/Tools/C_mk/Make.defs b/Tools/C_mk/Make.defs
index 88ab27f..60ac4fd 100644
--- a/Tools/C_mk/Make.defs
+++ b/Tools/C_mk/Make.defs
@@ -487,10 +487,24 @@ ifeq ($(COMP),Intel)
    endif

    ifeq ($(WHICHLINUX), CORI)
+     # This is a workaround for a known Cray bug which Cray does not plan to
+     # fix. The Intel compilers use GCC headers for C++11 but the Intel
+     # programming environment on the Crays does not find them without help. So
+     # we give it some help.
+     ifeq ($(USE_CXX11), TRUE)
+       ifneq ($(GCC_PATH), /opt/gcc/4.9.3)
+         $(error On Cori and Edison you must load the gcc/4.9.3 module to compile C++11 features with Intel compilers. This is a Cray bug)
+       endif
+     endif
      CXXOPTF += -g -O2 -ip -qopt-report=5 -qopt-report-phase=vec -fp-model source
    endif

    ifeq ($(WHICHLINUX), EDISON)
+     ifeq ($(USE_CXX11), TRUE)
+       ifneq ($(GCC_PATH), /opt/gcc/4.9.3)
+         $(error On Cori and Edison you must load the gcc/4.9.3 module to compile C++11 features with Intel compilers. This is a Cray bug)
+       endif
+     endif
      CXXOPTF += -g -O2 -ip -qopt-report=5 -qopt-report-phase=vec -fp-model source
    endif
```

**Comment by WeiqunZhang** [comment]
Looks good to me.

The reason I suggest to check the intel compiler version is that it used to
be for intel 14 one had to avoid 4.9.x.  So it might be the case intel 17
or 18 won't work with 4.9.3.  But I am probably worrying too much about the
future.

Weiqun

**Comment by bcfriesen** [comment]
I implemented a workaround for this issue in commit 870356bbba96593354337350bd8017ae4ce604dd.
